### ISSUE TITLE
Wrong size of the Embedding Matrix

### ISSUE THREAD

**Issue by alexiej** [fastai/fastai#1843]
**Describe the bug**
I’ve tried collaborative filtering, by example in the Google Colab (version 1.0.48, and newest library 1.0.50.post1)  from the ml-latest dataset (http://files.grouplens.org/datasets/movielens/ml-latest-small.zip). https://colab.research.google.com/drive/1qVwvgsMa1UMG5cqxZQu9r8zWznAsg2ay

```python
from fastai.collab import *
from fastai.tabular import *


#'http://files.grouplens.org/datasets/movielens/ml-latest-small.zip'
#path = untar_data(URLs.ML_SAMPLE)
#print(path.ls())

#!wget http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
#!unzip ml-latest-small.zip

ratings = pd.read_csv('ml-latest-small/ratings.csv')
all_movies = ratings['movieId'].unique().astype(str)
all_users = ratings['userId'].unique().astype(str)

print('Number of movies:', len(all_movies),'Number of users:', len(all_users))
ratings.head()

```
![image](https://user-images.githubusercontent.com/12593751/54677367-00af8a00-4b03-11e9-84cc-28c7ab223eee.png)

After splitting the data and create a data bunch number of classes and created Embedding model has a smaller number of items that I expected when I set `valid_pct` more than 0.

```python
user,title,rating = 'userId','movieId','rating'
data = CollabList.from_df(ratings, cat_names=[user,title],procs=Categorify)
data_split = data.split_by_rand_pct(valid_pct=0.2, seed=200).label_from_df(cols=rating)
print('classes: ', len(data_split.x.classes['movieId']))

data_bunch = data_split.databunch()
y_range = [0,5.5]
learn = collab_learner(data_bunch, n_factors=40, y_range=y_range, wd=1e-1)

print(len(learn.data.x.classes['movieId']))
print(learn.data.get_emb_szs())
learn.model
```
![image](https://user-images.githubusercontent.com/12593751/54677536-5f750380-4b03-11e9-82da-ce92ec5a5634.png)

![image](https://user-images.githubusercontent.com/12593751/54677512-53894180-4b03-11e9-9093-ef84b29fba90.png)

The problem is that size for Embedding Matrix are get only from the `train` dataset, but when some of the items are only in the `valid` dataset, they are not represented in the model.

**Provide your installation details**
```text
=== Software === 
python       : 3.6.7
fastai       : 1.0.50.post1
fastprogress : 0.1.20
torch        : 1.0.1
torch cuda   : None / is **Not available** 

=== Hardware === 
No GPUs available 

=== Environment === 
platform     : Windows-10-10.0.17763-SP0
conda env    : Unknown
python       : C:\PUBLIC\Anaconda3\python.exe
sys.path     : C:\PUBLIC\Sync\NAUKA\TYPORA\DATA-SCIENCE\fast.ai\collaborative
C:\PUBLIC\Sync\NAUKA\TYPORA\DATA-SCIENCE\fast.ai\collaborative
C:\PUBLIC\Anaconda3\python36.zip
C:\PUBLIC\Anaconda3\DLLs
C:\PUBLIC\Anaconda3\lib
C:\PUBLIC\Anaconda3
C:\Users\Arkadiusz\AppData\Roaming\Python\Python36\site-packages
C:\PUBLIC\Anaconda3\lib\site-packages
C:\PUBLIC\Anaconda3\lib\site-packages\Mako-1.0.7-py3.6.egg
C:\PUBLIC\Anaconda3\lib\site-packages\win32
C:\PUBLIC\Anaconda3\lib\site-packages\win32\lib
C:\PUBLIC\Anaconda3\lib\site-packages\Pythonwin
no supported gpus found on this system
```

**To Reproduce**
```
from fastai.collab import *
from fastai.tabular import *
import fastai
print(fastai.__version__)

#!wget http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
#!unzip ml-latest-small.zip

ratings = pd.read_csv('ml-latest-small/ratings.csv')
all_movies = ratings['movieId'].unique().astype(str)
all_users = ratings['userId'].unique().astype(str)

print('Number of movies:', len(all_movies),'Number of users:', len(all_users))

user,title,rating = 'userId','movieId','rating'
data = CollabList.from_df(ratings, cat_names=[user,title],procs=Categorify)
data_split = data.split_by_rand_pct(valid_pct=0.2, seed=200).label_from_df(cols=rating)
print('classes: ', len(data_split.x.classes['movieId']))

data_bunch = data_split.databunch()
y_range = [0,5.5]
learn = collab_learner(data_bunch, n_factors=40, y_range=y_range, wd=1e-1)

print(len(learn.data.x.classes['movieId']))
print(learn.data.get_emb_szs())
learn.model
```

https://colab.research.google.com/drive/1qVwvgsMa1UMG5cqxZQu9r8zWznAsg2ay

**Expected behaviour**

```text
classes:  9725
9725
[(611, 58), (9725, 262)]
EmbeddingDotBias(
  (u_weight): Embedding(611, 40)
  (i_weight): Embedding(9725, 40)
  (u_bias): Embedding(611, 1)
  (i_bias): Embedding(9725, 1)
)
```

**Screenshots**

![image](https://user-images.githubusercontent.com/12593751/54677017-533c7680-4b02-11e9-882f-1c2e1bddf843.png)
![image](https://user-images.githubusercontent.com/12593751/54677056-65b6b000-4b02-11e9-8a47-5343a245366c.png)

**Additional context**
My topic that describes the bug: https://forums.fast.ai/t/collaborative-filtering-different-size-of-embedding-matrix/41444

After debugging the code I found that in class `CollabList` in the function `split_by_list`, the class has the correct number of items in `inner_df`

```python
    def split_by_list(self, train, valid):
        "Split the data between `train` and `valid`."
        return self._split(self.path, train, valid)
```

![image](https://user-images.githubusercontent.com/12593751/54677795-ecb85800-4b03-11e9-96c6-c736fc811f7e.png)

but after that `ItemLists()` replace `inner_df` with only `test` DataFrame that is used to define `self.x` and next list of classes (`learn.data.x.classes['movieId']`), and size for the Embedding matrix (`self.get_emb_szs()`)

![image](https://user-images.githubusercontent.com/12593751/54677898-31dc8a00-4b04-11e9-9dfe-680e07397d8c.png)

**Comment by sgugger** [comment]
Hi there! Classes that are only present in the validation set as considered as unknown. This is the expected behavior and not a bug: if we put some embedding for those, they will stay in their random initial state since the model doesn't get any sample with those classes in the training set. It can't properly learn.

This is why you see the right number when setting a validation pct to 0. Another way to solve this would be to carefully compute validation indices so that all the classes are present in the training set and not use a random split. You would then use `split_by_idx` instead of `rand_split_by_pct`.
